cpp/ql/test/query-tests/Security/CWE/CWE-134/semmle/argv/argvLocal.c

```diff
@@ -168,4 +168,34 @@ int main(int argc, char **argv) {
 	int i10 = (int) argv[1];
 	printf((char *) i10);
 	printWrapper((char *) i10);
+
+	// BAD: b value comes from argv
+	{
+		char b[64];
+		char *bp = &b[0];
+		char *t;
+		if (0) {
+			t = 0;
+		} else {
+			t = bp;
+		}
+		memcpy(t, argv[1] + 1, 1);
+		printf(bp);
+		printWrapper(bp);
+	}
```

Comment:
Actually, I think the reason we don't catch these is because dataflow doesn't do a good job at alias analysis: We're never really good at catching flows like:
```cpp
// establish alias between p and q
// p = source();
// sink(q);
```
since dataflow only considers paths starting at `source` and onwards.

So that's probably why we don't catch the flow in this example.

With that said, these are some excellent testcases!

Reply:
They do work with def-use dataflow, so that's a bit "funny" to say the least.

Reply:
Yeah, I _think_ that's because the def-use dataflow does use the IR's alias analysis in some cases 🤔.

Reply:
I did not know that.

Reply:
(Let me double check this to make sure I'm not lying.)

Reply:
Yeah, there is a couple of cases of reliance on the IR alias analysis:
- In dataflow, when we're finding the input for a model that specifies indirect flow into the model: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L978 (and the same happens for taint flow).
- At function call boundaries we conflate a pointer and its indirection: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll#L98

Reply:
Should we have something similar in the case of use-use dataflow?

Reply:
It would certainly be good to make use of this information (hopefully in a more structured way than how we do it on `main` currently). https://github.com/github/codeql-c-team/issues/714 describes a(n old) version of this problem, but it should probably be updated to reflect the state of the use-use flow branch.